Keen/Dataset/Datasets.cs

```diff
@@ -0,0 +1,301 @@
+            }
+
+            var responseMessage = await _keenHttpClient
+                .GetAsync(datasetResultsUrl, _masterKey)
+                .ConfigureAwait( false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(false);
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinitionCollection>(responseString, GetSerializerSettings());
+        }
+
+        public async Task<IEnumerable<DatasetDefinition>> ListAllDefinitions()
+        {
+            var allDefinitions = new List<DatasetDefinition>();
+
+            var firstSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT);
+
+            if (firstSet?.Datasets == null)
+            {
+                throw new KeenException("Failed to fetch definition list");
+            }
+
+            if (!firstSet.Datasets.Any())
+            {
+                return allDefinitions;
+            }
+                
+            if (firstSet.Count <= firstSet.Datasets.Count())
+            {
+                return firstSet.Datasets;
+            }
+            
+            allDefinitions.AddRange(firstSet.Datasets);
+
+            do
+            {
+                var nextSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT, allDefinitions.Last().DatasetName);
+
+                if (nextSet?.Datasets == null || !nextSet.Datasets.Any())
+                {
+                    throw new KeenException("Failed to fetch definition list");
+                }
+
+                allDefinitions.AddRange(nextSet.Datasets);
+
+            } while (firstSet.Count > allDefinitions.Count);
+
+            return allDefinitions;
+        }
+
+        public async Task DeleteDataset(string datasetName)
+        {
+            if (string.IsNullOrWhiteSpace(datasetName))
+            {
+                throw new KeenException("A dataset name is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var responseMessage = await _keenHttpClient
+                .DeleteAsync(this.GetDatasetUrl(datasetName), _masterKey)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            if (responseMessage.StatusCode == HttpStatusCode.NoContent)
```

Comment:
This type of info, i.e. the knowledge that we expect a 204 when we DELETE cached queries/datasets, would also be handy to have in another layer, which is covered by Issue #55.